rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+    bypass its outputs entirely and only use the flake mechanism to
+    get its source tree.
+
+  * `rev`: The commit hash of the flake's Git repository.
+
+  * `revCount`: The number of ancestors of the revision `rev`. This is
+    not available for `github:...` repositories (see below), since
+    they're fetched as tarballs rather than as Git repositories.
+
+  * `lastModified`: The commit time of the revision `rev`, in the
+    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
+    this is available for both Git and GitHub repositories, so it's
+    useful for generating (hopefully) monotonically increasing version
+    strings.
+
+  The special input named `self` refers to the outputs and source tree
+  of *this* flake.
+
+  A number of outputs have a special meaning, as discussed below.
+
+## Well-known outputs
+
+A number of outputs have a specific meaning to Nix or other tools like
+Hydra. Currently, these are:
+
+* `packages`: A set of derivations used as a default by most `nix`
+  commands. For example, `nix run nixpkgs:hello` uses the
+  `packages.hello` attribute of the `nixpkgs` flake. It cannot contain
+  any non-derivation attributes. This also means it cannot be a nested
+  set! (The rationale is that supporting nested sets requires Nix to
+  evaluate each attribute in the set, just to discover which packages
+  are provided.)
+
+* `defaultPackage`: A derivation used as a default by most `nix`
+  commands if no attribute is specified. For example, `nix run
+  dwarffs` uses the `defaultPackage` attribute of the `dwarffs` flake.
+
+* `checks`: A non-nested set of derivations built by the `nix flake
+  check` command, or by Hydra if a flake does not have a `hydraJobs`
+  attribute.
+
+* `hydraJobs`: A nested set of derivations built by Hydra.
+
+* `devShell`: A derivation that defines the shell environment used by
+  `nix dev-shell` if no specific attribute is given. If it does not
+  exist, then `nix dev-shell` will use `defaultPackage`.
+
+* `apps`: A set of app definitions used by the `nix app` command. For
+  example, `nix app blender-bin:blender_2_79` uses the
+  `apps.blender_2_79` output of the `blender-bin` flake.
+
+* `defaultApp`: An app definition used by the `nix app` command when no
+  specific attribute is given. For example, `nix app blender-bin` uses
+  the `defaultApp` output of the `blender-bin` flake.
+
+TODO: NixOS-related outputs such as `nixosModules` and `nixosSystems`.
+
+## Flake references
+
+Flake references are a vaguely URL-like syntax to specify which flake
+to use. This is used on the command line (e.g. in `nix build
+nixpkgs:hello`, `nixpkgs` is a flake reference), and in the list of
+flake dependencies in `flake.nix` (e.g. in `inputs = [ "nixpkgs" ];`).
+
+Currently the following types of flake references are supported:
+
+* Git repositories. These have the form
+
+      (http|https|ssh|git|file):(//<server>)?<path>(\?<params>)?
```

Comment:
```suggestion
      (git+http|git+https|git+ssh|git|git+file):(//<server>)?<path>(\?<params>)?
```
It might be useful to use a prefix to avoid confusion in the future on the method of access.